### PR TITLE
feat(astro): Add support for View Transitions

### PR DESCRIPTION
**Pull request by wobsoriano**
## Description

This PR adds [Astro View Transitions](https://docs.astro.build/en/guides/view-transitions/) support to our Clerk components. It is basically the "SPA mode" for Astro.

Before this change, Clerk components didn't work properly with Astro View Transitions. When using the `<ViewTransitions />` component in an Astro app, our Clerk components weren't remounting during page navigation.

The solution was to:

1. Run the [`runInjectionScript`](https://github.com/clerk/javascript/blob/3b50b67bd40da33c9e36773aa05462717e9f44cc/packages/astro/src/internal/create-clerk-instance.ts#L32) function in the [`astro:page-load`](https://docs.astro.build/en/guides/view-transitions/#astropage-load) to ensure Clerk components are properly mounted when switching pages.
2. Create a [custom swap function](https://docs.astro.build/en/guides/view-transitions/#building-a-custom-swap-function) and keep the elements added by `@emotion/cache`. This maintains the styling across page transitions.

Here's a basic demo using `<SignIn />`, `<UserProfile />` and `<UserButton />` components being rendered in a view transitions enabled pages:

https://github.com/user-attachments/assets/16086c0f-34c2-48ac-91f8-eb8ad84b8c53

Resolves ECO-64

## Checklist

- [x] `npm test` runs as expected.
- [x] `npm run build` runs as expected.
- [x] (If applicable) [JSDoc comments](https://jsdoc.app/about-getting-started.html) have been added or updated for any package exports
- [ ] (If applicable) [Documentation](https://github.com/clerk/clerk-docs) has been updated

## Type of change

- [ ] 🐛 Bug fix
- [x] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

### CODE REVIEW

File: packages/astro/src/integration/create-integration.ts [hunk 1]
```diff
@@ -109,8 +109,26 @@ function createIntegration<Params extends HotloadAstroClerkIntegrationParams>()
...
+                const clonedEl = clerkComponents.cloneNode(true);
+                e.newDocument.body.appendChild(clonedEl);
+              }
+
+              e.swap = () => swapDocument(e.newDocument);
+            });
+
+            document.addEventListener('astro:page-load', async (e) => {
+              await runInjectionScript(${JSON.stringify(internalParams)});
+            })`,
```
Comment (wobsoriano) [inline packages/astro/src/integration/create-integration.ts:131]: Both of these events will only run when view transitions is enabled

File: packages/astro/src/internal/swap-document.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+ *
+ * See https://docs.astro.build/en/guides/view-transitions/#building-a-custom-swap-function
+ */
+export function swapDocument(doc: Document) {
+  swapFunctions.deselectScripts(doc);
+  swapFunctions.swapRootAttributes(doc);
+
+  // Keep the elements created by `@emotion/cache`
+  const emotionElements = document.querySelectorAll(`style[${EMOTION_ATTR}]`);
+  swapHeadElements(doc, Array.from(emotionElements));
```
Comment (wobsoriano) [inline packages/astro/src/internal/swap-document.ts:20]: We need to keep the `<style>` element added by `@emotion/cache` to keep the styling when switching between pages.

Screenshot 2024-10-17 at 1 34 41 PM

The default Astro swap function will remove this one

File: packages/astro/src/internal/swap-document.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
+// eslint-disable-next-line import/no-unresolved
+import { swapFunctions } from 'astro:transitions/client';
```
Comment (panteliselef) [inline packages/astro/src/internal/swap-document.ts:2]: ❓ This is always bundled with Astro ? Is there a case where the module will not exists causing our integration to break ?

**Review by panteliselef — APPROVED** [review]
Seems great. In the video, i see that the "afterSignIn" navigation did not run with the page transitions (fade-in/out). Is this expected ?

File: packages/astro/src/internal/swap-document.ts (hunk 3 above)
Comment (wobsoriano) [reply packages/astro/src/internal/swap-document.ts:2]: It is indeed bundled with Astro (`astro/virtual-modules/transitions-swap-functions.js`), but is not available in Astro v3

**Comment by clerk-cookie** [comment]
Hey @wobsoriano - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/astro | 1.4.0-snapshot.va2afb95 |
| @clerk/backend | 1.14.2-snapshot.va2afb95 |
| @clerk/chrome-extension | 1.3.21-snapshot.va2afb95 |
| @clerk/clerk-js | 5.28.0-snapshot.va2afb95 |
| @clerk/elements | 0.17.0-snapshot.va2afb95 |
| @clerk/clerk-expo | 2.2.27-snapshot.va2afb95 |
| @clerk/express | 1.3.2-snapshot.va2afb95 |
| @clerk/fastify | 2.0.4-snapshot.va2afb95 |
| @clerk/localizations | 3.3.1-snapshot.va2afb95 |
| @clerk/nextjs | 5.8.0-snapshot.va2afb95 |
| @clerk/clerk-react | 5.12.1-snapshot.va2afb95 |
| @clerk/remix | 4.2.40-snapshot.va2afb95 |
| @clerk/clerk-sdk-node | 5.0.53-snapshot.va2afb95 |
| @clerk/shared | 2.10.0-snapshot.va2afb95 |
| @clerk/tanstack-start | 0.4.16-snapshot.va2afb95 |
| @clerk/testing | 1.3.14-snapshot.va2afb95 |
| @clerk/themes | 2.1.38-snapshot.va2afb95 |
| @clerk/types | 4.27.0-snapshot.va2afb95 |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/astro`
```sh
npm i @clerk/astro@1.4.0-snapshot.va2afb95 --save-exact
```
`@clerk/backend`
```sh
npm i @clerk/backend@1.14.2-snapshot.va2afb95 --save-exact
```
`@clerk/chrome-extension`
```sh
npm i @clerk/chrome-extension@1.3.21-snapshot.va2afb95 --save-exact
```
`@clerk/clerk-js`
```sh
npm i @clerk/clerk-js@5.28.0-snapshot.va2afb95 --save-exact
```
`@clerk/elements`
```sh
npm i @clerk/elements@0.17.0-snapshot.va2afb95 --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@2.2.27-snapshot.va2afb95 --save-exact
```
`@clerk/express`
```sh
npm i @clerk/express@1.3.2-snapshot.va2afb95 --save-exact
```
`@clerk/fastify`
```sh
npm i @clerk/fastify@2.0.4-snapshot.va2afb95 --save-exact
```
`@clerk/localizations`
```sh
npm i @clerk/localizations@3.3.1-snapshot.va2afb95 --save-exact
```
`@clerk/nextjs`
```sh
npm i @clerk/nextjs@5.8.0-snapshot.va2afb95 --save-exact
```
`@clerk/clerk-react`
```sh
npm i @clerk/clerk-react@5.12.1-snapshot.va2afb95 --save-exact
```
`@clerk/remix`
```sh
npm i @clerk/remix@4.2.40-snapshot.va2afb95 --save-exact
```
`@clerk/clerk-sdk-node`
```sh
npm i @clerk/clerk-sdk-node@5.0.53-snapshot.va2afb95 --save-exact
```
`@clerk/shared`
```sh
npm i @clerk/shared@2.10.0-snapshot.va2afb95 --save-exact
```
`@clerk/tanstack-start`
```sh
npm i @clerk/tanstack-start@0.4.16-snapshot.va2afb95 --save-exact
```
`@clerk/testing`
```sh
npm i @clerk/testing@1.3.14-snapshot.va2afb95 --save-exact
```
`@clerk/themes`
```sh
npm i @clerk/themes@2.1.38-snapshot.va2afb95 --save-exact
```
`@clerk/types`
```sh
npm i @clerk/types@4.27.0-snapshot.va2afb95 --save-exact
```

**Comment by wobsoriano** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @wobsoriano - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/astro | 1.4.0-snapshot.vb58e76b |
| @clerk/backend | 1.14.2-snapshot.vb58e76b |
| @clerk/chrome-extension | 1.3.21-snapshot.vb58e76b |
| @clerk/clerk-js | 5.28.0-snapshot.vb58e76b |
| @clerk/elements | 0.17.0-snapshot.vb58e76b |
| @clerk/clerk-expo | 2.2.27-snapshot.vb58e76b |
| @clerk/express | 1.3.2-snapshot.vb58e76b |
| @clerk/fastify | 2.0.4-snapshot.vb58e76b |
| @clerk/localizations | 3.3.1-snapshot.vb58e76b |
| @clerk/nextjs | 5.8.0-snapshot.vb58e76b |
| @clerk/clerk-react | 5.12.1-snapshot.vb58e76b |
| @clerk/remix | 4.2.40-snapshot.vb58e76b |
| @clerk/clerk-sdk-node | 5.0.53-snapshot.vb58e76b |
| @clerk/shared | 2.10.0-snapshot.vb58e76b |
| @clerk/tanstack-start | 0.4.16-snapshot.vb58e76b |
| @clerk/testing | 1.3.14-snapshot.vb58e76b |
| @clerk/themes | 2.1.38-snapshot.vb58e76b |
| @clerk/types | 4.27.0-snapshot.vb58e76b |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/astro`
```sh
npm i @clerk/astro@1.4.0-snapshot.vb58e76b --save-exact
```
`@clerk/backend`
```sh
npm i @clerk/backend@1.14.2-snapshot.vb58e76b --save-exact
```
`@clerk/chrome-extension`
```sh
npm i @clerk/chrome-extension@1.3.21-snapshot.vb58e76b --save-exact
```
`@clerk/clerk-js`
```sh
npm i @clerk/clerk-js@5.28.0-snapshot.vb58e76b --save-exact
```
`@clerk/elements`
```sh
npm i @clerk/elements@0.17.0-snapshot.vb58e76b --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@2.2.27-snapshot.vb58e76b --save-exact
```
`@clerk/express`
```sh
npm i @clerk/express@1.3.2-snapshot.vb58e76b --save-exact
```
`@clerk/fastify`
```sh
npm i @clerk/fastify@2.0.4-snapshot.vb58e76b --save-exact
```
`@clerk/localizations`
```sh
npm i @clerk/localizations@3.3.1-snapshot.vb58e76b --save-exact
```
`@clerk/nextjs`
```sh
npm i @clerk/nextjs@5.8.0-snapshot.vb58e76b --save-exact
```
`@clerk/clerk-react`
```sh
npm i @clerk/clerk-react@5.12.1-snapshot.vb58e76b --save-exact
```
`@clerk/remix`
```sh
npm i @clerk/remix@4.2.40-snapshot.vb58e76b --save-exact
```
`@clerk/clerk-sdk-node`
```sh
npm i @clerk/clerk-sdk-node@5.0.53-snapshot.vb58e76b --save-exact
```
`@clerk/shared`
```sh
npm i @clerk/shared@2.10.0-snapshot.vb58e76b --save-exact
```
`@clerk/tanstack-start`
```sh
npm i @clerk/tanstack-start@0.4.16-snapshot.vb58e76b --save-exact
```
`@clerk/testing`
```sh
npm i @clerk/testing@1.3.14-snapshot.vb58e76b --save-exact
```
`@clerk/themes`
```sh
npm i @clerk/themes@2.1.38-snapshot.vb58e76b --save-exact
```
`@clerk/types`
```sh
npm i @clerk/types@4.27.0-snapshot.vb58e76b --save-exact
```

**Comment by mlafeldt** [comment]
Hey @wobsoriano!

I'm using @clerk/astro v1.4.7, which should contain this change. However, I'm still seeing exactly this behavior on Cloudflare Pages with the UserButton disappearing randomly on navigation:

> Before this change, Clerk components didn't work properly with Astro View Transitions. When using the `<ViewTransitions />` component in an Astro app, our Clerk components weren't remounting during page navigation.

My workaround is to use UserButton from `@clerk/astro/react` instead of `@clerk/astro/components`, which works fine.

Maybe they changed something in the latest Astro release?

```
❯ bun astro info
Astro                    v4.16.9
Node                     v22.9.0
System                   macOS (arm64)
Package Manager          bun
Output                   hybrid
Adapter                  @astrojs/cloudflare
Integrations             @clerk/astro/integration
                         @astrojs/react
                         @astrojs/tailwind
```

**Comment by wobsoriano** [comment]
@mlafeldt that is interesting. I will try and update my astro template [here](https://astro-clerk-template.vercel.app/) and check if the UI components from `@clerk/astro/components` fails 👀

**Comment by mlafeldt** [comment]
@wobsoriano It looks like the UserButton disappears when navigating between pre-rendered and server-rendered pages. One page in my app uses React without SSR. Adding `export const prerender = false` to that page fixes the problem.

**Pull request merged**